Generate Client using OpenAPI spec, Update Setup Guide and Quickstart (#2)
ballerina-platform/module-ballerinax-hubspot.marketing.emails: ballerina/Package.md

## Overview
[HubSpot](https://www.hubspot.com) is an AI-powered customer relationship management (CRM) platform. 

> The `ballerinax/hubspot.marketing.emails` package offers APIs to connect and interact with [HubSpot Marketing Emails API](https://developers.hubspot.com/docs/reference/api/marketing/emails/marketing-emails) endpoints, specifically based on [OpenAPI specification for HubSpot Marketing Emails API v3](https://github.com/HubSpot/HubSpot-public-api-spec-collection/blob/main/PublicApiSpecs/Marketing/Marketing%20Emails/Rollouts/145892/v3/marketingEmails.json).

Reviewer: Check the comment on the README file and update the same here.

    ```ballerina 
    configurable string clientId = ?;
    configurable string clientSecret = ?;
    configurable string refreshToken = ?;

    final hsmemails:ConnectionConfig hsmeventsConfig = {
        auth : {
            clientId,
            clientSecret,
            refreshToken,
            credentialBearer: oauth2:POST_BODY_BEARER
        }
    };

    final hsmemails:Client hsmevents = check new (hsmeventsConfig, "https://api.hubapi.com/marketing/v3/emails");
    ```

### Step 3: Invoke the connector operation

Now, utilize the available connector operations. A sample usecase is shown below.

#### Create a Marketing Event
    
```ballerina
public function main() returns error? {
    hsmemails:AggregateEmailStatistics emailStatistics = check hubspotClient->/marketing/v3/emails/statistics/list({}, 
     {
         startTimestamp: "2024-12-12T04:27:02Z",
         endTimestamp: "2024-12-19T04:27:02Z"
     });
}
```

## Examples

The `Hubspot Marketing Emails` connector provides practical examples illustrating usage in various scenarios. Explore these examples, covering use cases:

> 1. [Bulk Change Reply To Email](./BulkChangeReplyEmail/) - Change the Reply To and Custom Reply To email address of all draft emails

Reviewer: ```suggestion
1. [Bulk Change Reply To Email](../examples/BulkChangeReplyEmail/) - Change the Reply To and Custom Reply To email address of all draft emails
```

Using this API, users can develop applications easily that enables you to track marketing emails.

## Setup guide

To use the HubSpot Marketing Events connector, you must have access to the HubSpot API through a HubSpot developer account and a HubSpot App under it. Therefore you need to register for a developer account at HubSpot if you don't have one already.

### Step 1: Create/Login to a HubSpot Developer Account

If you have an account already, go to the [HubSpot developer portal](https://app.hubspot.com/)

If you don't have a HubSpot Developer Account you can sign up to a free account [here](https://developers.hubspot.com/get-started)

### Step 2 (Optional): Create a [Developer Test Account](https://developers.hubspot.com/beta-docs/getting-started/account-types#developer-test-accounts) under your account

Within app developer accounts, you can create developer test accounts to test apps and integrations without affecting any real HubSpot data.

> **Note:** These accounts are only for development and testing purposes. In production you should not use Developer Test Accounts.

Reviewer: ```suggestion
> **Note:** These accounts are only for development and testing purposes. In production you should not use Developer Test Accounts.
```

1. Go to Test Account section from the left sidebar.

>    ![Hubspot Developer Portal](../docs/setup/resources/test_acc_1.png)

Reviewer: ```suggestion
   ![Hubspot Developer Portal](https://raw.githubusercontent.com/ballerina-platform/module-ballerinax-hubspot.marketing.emails/main/docs/setup/resources/test_acc_1.png)
```

> 

Reviewer: ```suggestion
```

2. Click Create developer test account.

>    ![Hubspot Developer Test Account](../docs/setup/resources/test_acc_2.png)

Reviewer: ```suggestion
   ![Hubspot Developer Test Account](https://raw.githubusercontent.com/ballerina-platform/module-ballerinax-hubspot.marketing.emails/main/docs/setup/resources/test_acc_2.png)
```

3. In the dialogue box, give a name to your test account and click create.

>    ![Hubspot Developer Test Account](../docs/setup/resources/test_acc_3.png)

Reviewer: ```suggestion
   ![Hubspot Developer Test Account](https://raw.githubusercontent.com/ballerina-platform/module-ballerinax-hubspot.marketing.emails/main/docs/setup/resources/test_acc_3.png)
```

### Step 3: Create a HubSpot App under your account.

1. In your developer account, navigate to the "Apps" section. Click on "Create App"

>    ![Hubspot Create App](../docs/setup/resources/create_app_1.png )

Reviewer: ```suggestion
   ![Hubspot Create App](https://raw.githubusercontent.com/ballerina-platform/module-ballerinax-hubspot.marketing.emails/main/docs/setup/resources/create_app_1.png )
```

2. Provide the necessary details, including the app name and description.

### Step 4: Configure the Authentication Flow.

1. Move to the Auth Tab. (Second tab next to App Info)

>    ![Hubspot Developer Config Auth](../docs/setup/resources/auth_section.png )

Reviewer: ```suggestion
   ![Hubspot Developer Config Auth](https://raw.githubusercontent.com/ballerina-platform/module-ballerinax-hubspot.marketing.emails/main/docs/setup/resources/auth_section.png )
```

2. In the Scopes section, add the following scope for your app using the "Add new scope" button.

   `content`

>    ![Hubspot Developer App Add Scopes](../docs/setup/resources/scopes.png )

Reviewer: ```suggestion
   ![Hubspot Developer App Add Scopes](https://raw.githubusercontent.com/ballerina-platform/module-ballerinax-hubspot.marketing.emails/main/docs/setup/resources/scopes.png )
```

4. Add your Redirect URI in the relevant section. You can also use `localhost` addresses for local development purposes. Click Create App.

>    ![Hubspot Create Developer App](../docs/setup/resources/create_app_final.png )

Reviewer: ```suggestion
   ![Hubspot Create Developer App](https://raw.githubusercontent.com/ballerina-platform/module-ballerinax-hubspot.marketing.emails/main/docs/setup/resources/create_app_final.png )
```

>    `content`

Reviewer: ```suggestion
   * `content`
```

### Step 5: Get your Client ID and Client Secret

- Navigate to the Auth section of your app. Make sure to save the provided Client ID and Client Secret.

>    ![Hubspot Get Credentials](../docs/setup/resources/get_credentials.png )

Reviewer: ```suggestion
   ![Hubspot Get Credentials](https://raw.githubusercontent.com/ballerina-platform/module-ballerinax-hubspot.marketing.emails/main/docs/setup/resources/get_credentials.png )
```

### Step 6: Setup Authentication Flow

Before proceeding with the Quickstart, ensure you have obtained the Access Token using the following steps:

1. Create an authorization URL using the following format:

   ```
   https://app.hubspot.com/oauth/authorize?client_id=<YOUR_CLIENT_ID>&scope=<YOUR_SCOPES>&redirect_uri=<YOUR_REDIRECT_URI>
   ```

   Replace the `<YOUR_CLIENT_ID>`, `<YOUR_REDIRECT_URI>` and `<YOUR_SCOPES>` with your specific value.

> **NOTE:** If you are using a `localhost` redirect url, make sure to have a listener running at the relevant port before executing the next step.

2. Paste it in the browser and select your developer test account to install the app when prompted.

>    ![Hubspot Get Auth Code](../docs/setup/resources/install_app.png)

Reviewer: ```suggestion
   ![Hubspot Get Auth Code](https://raw.githubusercontent.com/ballerina-platform/module-ballerinax-hubspot.marketing.emails/main/docs/setup/resources/install_app.png)
```

     "expires_in": 1800
   }
   ```

5. Store the access token securely for use in your application.

## Quickstart

To use the `HubSpot Marketing Emails` connector in your Ballerina application, update the `.bal` file as follows:

### Step 1: Import the module

Import the `hubspot.marketing.emails` module and `oauth2` module.

```ballerina
import ballerinax/hubspot.marketing.emails as hsmemails;
import ballerina/oauth2;
```

### Step 2: Instantiate a new connector

1. Create a `Config.toml` file and, configure the obtained credentials in the above steps as follows:

   ```toml
    clientId = <Client Id>
    clientSecret = <Client Secret>
    refreshToken = <Refresh Token>
   ```

2. Instantiate a `hsmevents:ConnectionConfig` with the obtained credentials and initialize the connector with it.

    ```ballerina 
    configurable string clientId = ?;
    configurable string clientSecret = ?;
    configurable string refreshToken = ?;

    final hsmemails:ConnectionConfig hsmeventsConfig = {
        auth : {
            clientId,
            clientSecret,
            refreshToken,
            credentialBearer: oauth2:POST_BODY_BEARER
        }
    };

>     final hsmemails:Client hsmevents = check new (hsmeventsConfig, "https://api.hubapi.com/marketing/v3/emails");

Reviewer: ```suggestion
    final hsmemails:Client hsmevents = check new (hsmeventsConfig);
```

    ```

### Step 3: Invoke the connector operation

Now, utilize the available connector operations. A sample usecase is shown below.

#### Get statistics for Marketing Emails
    
```ballerina
public function main() returns error? {
    hsmemails:AggregateEmailStatistics emailStatistics = check hubspotClient->/marketing/v3/emails/statistics/list({}, 
     {
         startTimestamp: "2024-12-12T04:27:02Z",
         endTimestamp: "2024-12-19T04:27:02Z"
     });
}
```

## Examples

The `Hubspot Marketing Emails` connector provides practical examples illustrating usage in various scenarios. Explore these examples, covering use cases:

1. [Bulk Change Reply To Email](https://raw.githubusercontent.com/ballerina-platform/module-ballerinax-hubspot.marketing.emails/main/examples/bulk_change_reply_email/) - Change the Reply To and Custom Reply To email address of all draft emails

> 2. [Marketing Email Statistics Logger](https://raw.githubusercontent.com/ballerina-platform/module-ballerinax-hubspot.marketing.emails/main/examples/email_stat_logger/) - Retrieve and log the statistics of marketing emails during a specific time period.

Reviewer: Update the links as suggested in the `Module.md`

## Overview

> [//]: # "TODO: Add overview mentioning the purpose of the module, supported REST API versions, and other high-level details."

Reviewer: Let's remove the "TODO" parts here.

> [HubSpot](https://www.hubspot.com/our-story) is a customer platform with all the software, integrations, and resources users need to connect thier marketing, sales, and customer service.

Reviewer: Let's follow the provided template for the introduction part.

> The `ballerinax/hubspot.marketing.emails` package offers APIs to connect and interact with [HubSpot API for Marketing Emails](https://developers.hubspot.com/docs/reference/api/marketing/emails/marketing-emails) endpoints, specifically based on [HubSpot Marketing Emails API v3 OpenAPI spec](https://github.com/HubSpot/HubSpot-public-api-spec-collection/blob/main/PublicApiSpecs/Marketing/Marketing%20Emails/Rollouts/145892/v3/marketingEmails.json).

Reviewer 2: ```suggestion
The `ballerinax/hubspot.marketing.emails` package offers APIs to connect and interact with [HubSpot Marketing Emails API](https://developers.hubspot.com/docs/reference/api/marketing/emails/marketing-emails) endpoints, specifically based on [OpenAPI specification for HubSpot Marketing Emails API v3](https://github.com/HubSpot/HubSpot-public-api-spec-collection/blob/main/PublicApiSpecs/Marketing/Marketing%20Emails/Rollouts/145892/v3/marketingEmails.json).
```

Using this API, users can develop applications easily that enables you to track marketing emails.

## Setup guide

[//]: # "TODO: Add detailed steps to obtain credentials and configure the module."

To use the HubSpot Marketing Events connector, you must have access to the HubSpot API through a HubSpot developer account and a HubSpot App under it. Therefore you need to register for a developer account at HubSpot if you don't have one already.

### Step 1: Create/Login to a HubSpot Developer Account

If you have an account already, go to the [HubSpot developer portal](https://app.hubspot.com/)

If you don't have a HubSpot Developer Account you can sign up to a free account [here](https://developers.hubspot.com/get-started)

### Step 2 (Optional): Create a [Developer Test Account](https://developers.hubspot.com/beta-docs/getting-started/account-types#developer-test-accounts) under your account

Within app developer accounts, you can create developer test accounts to test apps and integrations without affecting any real HubSpot data.

> **_These accounts are only for development and testing purposes. In production you should not use Developer Test Accounts._**

Reviewer: ```suggestion
> **Note:** These accounts are only for development and testing purposes. In production you should not use Developer Test Accounts.**
```

1. Go to Test Account section from the left sidebar.

>    <img src=../docs/setup/resources/test_acc_1.png alt="Hubspot developer portal" style="width: 70%;">

Reviewer: Let's use the markdown format here. Please check all the places.

```suggestion
   ![Hubspot developer portal](../docs/setup/resources/test_acc_1.png)
```

2. Click Create developer test account.

   <img src=../docs/setup/resources/test_acc_2.png alt="Hubspot developer testacc" style="width: 70%;">

3. In the dialogue box, give a name to your test account and click create.

   <img src=../docs/setup/resources/test_acc_3.png alt="Hubspot developer testacc3" style="width: 70%;">

### Step 3: Create a HubSpot App under your account.

1. In your developer account, navigate to the "Apps" section. Click on "Create App"

   <img src=../docs/setup/resources/create_app_1.png alt="Hubspot app creation 1 testacc3" style="width: 70%;">

2. Provide the necessary details, including the app name and description.

### Step 4: Configure the Authentication Flow.

1. Move to the Auth Tab. (Second tab next to App Info)

> 

Reviewer: Remove extra empty lines. Check similar places.

```suggestion
```

   <img src=../docs/setup/resources/auth_section.png alt="Hubspot app creation 2 testacc3" style="width: 70%;">

2. In the Scopes section, add the following scope for your app using the "Add new scope" button.

   `content`

   <img src=../docs/setup/resources/scopes.png alt="Hubspot app creation 1 testacc3" style="width: 70%;">

> 4. Add your Redirect URI in the relevant section. You can also use localhost addresses for local development purposes. Click Create App.

Reviewer: ```suggestion
4. Add your Redirect URI in the relevant section. You can also use `localhost` addresses for local development purposes. Click Create App.
```

   <img src=../docs/setup/resources/create_app_final.png alt="Hubspot app creation 1 testacc3" style="width: 70%;">

### Step 5: Get your Client ID and Client Secret

- Navigate to the Auth section of your app. Make sure to save the provided Client ID and Client Secret.

   <img src=../docs/setup/resources/get_credentials.png alt="Hubspot app creation 1 testacc3" style="width: 70%;">

### Step 6: Setup Authentication Flow

Before proceeding with the Quickstart, ensure you have obtained the Access Token using the following steps:

1. Create an authorization URL using the following format:

   ```
   https://app.hubspot.com/oauth/authorize?client_id=<YOUR_CLIENT_ID>&scope=<YOUR_SCOPES>&redirect_uri=<YOUR_REDIRECT_URI>
   ```

   Replace the `<YOUR_CLIENT_ID>`, `<YOUR_REDIRECT_URI>` and `<YOUR_SCOPES>` with your specific value.

>     **_NOTE: If you are using a localhost redirect url, make sure to have a listner running at the relevant port before executing the next step. You can use [this gist](https://gist.github.com/lnash94/0af47bfcb7cc1e3d59e06364b3c86b59) and run it using `bal run`. Alternatively, you can use any other method to bind a listner to the port._**

Reviewer: Let's not use the link to the gist here.

```suggestion
> **NOTE:** If you are using a `localhost` redirect url, make sure to have a listener running at the relevant port before executing the next step.
```

2. Paste it in the browser and select your developer test account to intall the app when prompted.

   <img src=../docs/setup/resources/install_app.png alt="Hubspot app creation 1 testacc3" style="width: 70%;">

3. A code will be displayed in the browser. Copy the code.

> 
>    ```
>    Received code: na1-129d-860c-xxxx-xxxx-xxxxxxxxxxxx
>    ```

Reviewer: We don't need this part.
```suggestion
```
